### PR TITLE
Redesign file switcher (a la Sublime Text)

### LINKED ISSUE #2141: File list management broken in master
_From [christop...@gmail.com](https://code.google.com/u/105753816054647931594/) on 2015-01-26T11:31:23Z_

Spyder Version:  2.4.0dev
Python Version:  2.7.6
Qt Version    :  4.8.6, PyQt4 (API v2) 4.11.3 on Darwin
pyflakes >=0.5.0:  0.5.0 (OK)
pep8 >=0.6      :  None (NOK)
IPython >=1.0   :  3.0.0-dev (OK)
zmq >=2.1.11    :  14.0.1 (OK)
pygments >=1.6  :  1.6 (OK)
pandas >=0.13.1 :  None (NOK)
sphinx >=0.6.6  :  1.2.3 (OK)
psutil >=0.3    :  0.7.1 (OK)
jedi >=0.8.1    :  None (NOK)
rope >=0.9.2    :  0.9.4 (OK)
matplotlib >=1.0:  1.4.0 (OK)
sympy >=0.7.0   :  0.7.2 (OK)
pylint >=0.25   :  None (NOK) 

What steps will reproduce the problem?

 1. Hit Ctrl-e in editor 

What is the expected output? What do you see instead?

 Open 'File list management' dialog. I get a stacktrace on the console instead.

Please provide any additional information below

.
my fix is the following:

diff -r 358150850acf spyderlib/widgets/editor.py
--- a/spyderlib/widgets/editor.py	Mon Jan 26 17:04:29 2015 +0100
+++ b/spyderlib/widgets/editor.py	Mon Jan 26 17:23:08 2015 +0100
@@ -74,7 +74,7 @@
 
         flabel = QLabel(_("Filter:"))
         self.edit = QLineEdit(self)
-        self.edit.returnPressed.connect(self.edit_file)
+        self.edit.returnPressed.connect(self.edit_file_slot)
         self.edit.textChanged.connect(lambda text: self.synchronize(0))
         fhint = QLabel(_("(press \<b>Enter</b> to edit file)"))
         edit_layout = QHBoxLayout()
@@ -85,12 +85,12 @@
         self.listwidget = QListWidget(self)
         self.listwidget.setResizeMode(QListWidget.Adjust)
         self.listwidget.itemSelectionChanged.connect(self.item_selection_changed)
-        self.listwidget.itemActivated.connect(self.edit_file)
+        self.listwidget.itemActivated.connect(self.edit_file_slot)
 
         btn_layout = QHBoxLayout()
         edit_btn = create_toolbutton(self, icon=get_icon('edit.png'),
                      text=_("&Edit file"), autoraise=False,
-                     triggered=self.edit_file, text_beside_icon=True)
+                     triggered=self.edit_file_slot, text_beside_icon=True)
         edit_btn.setMinimumHeight(28)
         btn_layout.addWidget(edit_btn)
 
@@ -121,7 +121,7 @@
         self.buttons = (edit_btn, close_btn)
 
     @-Slot()
-    def edit_file(self):
+    def edit_file_slot(self):
         row = self.listwidget.currentRow()
         if self.listwidget.count() > 0 and row >= 0:
             self.edit_file.emit(self.indexes[row])

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=2141_

### LINKED ISSUE #2340: SublimeText-like file switching widget
In Sublime Text, you press Ctrl-P to get an intelligent, fast switching between open files/files in the current folder.  See [here](http://sublime-text-unofficial-documentation.readthedocs.org/en/latest/file_management/file_management.html).

I find myself trying to use this shortcut in Spyder, and then get annoyed when I have to click through little buttons at the top of the editor.

I imagine that this tool is reasonably straightforward to implement as it's fairly stand alone.

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #2340 
Fixes #2141

---
## Description
Continues where PR #2343 left.

The file switcher now looks like this:

### File list
switcher-symbol

### Symbol list 
switcher-symbol2

## Original Features
- [x] matching text is highlighted in bold
- [x] up/down keys can be used to navigate list when text field has focus
- [x] navigating the list changes the active tab in the editor, but will revert if dialog is canceled
- [x] text field has focus when opened
- [x] the dialog is more compact and is positioned at the top, centre, of the editor
- [x] global context shortcut - I couldn't work out how to change this and keep Spyder happy.  I'd suggest Ctrl-P as default, but the main thing is to make it global rather than Editor-specific.
- [x] path info in the list in addition to file names (I think this is relatively easy now that there's an HTMLdelegate involved, but I could be wrong).
- [x] goto line - there is currently a separate dialog for this, but it might be nice to combine the two as in Sublime.

## New Features
- [x] Code leanup and refactoring
- [x] fuzzy matching
- [x] dialog resize based on contents
- [x] goto symbol - this would basically use the same data as the Outline tool. Again, this is something provided by Sublime.
- [x] Fix line number displayed
- [x] Restore original line numbers on cancel when looking for symbols
- [x] Fill-in missing docstring

## Todo
- [x] Fix py3 compatibility
- [x] Fix bug with cells in outline data

## For another PR
- Improve contrast on OSX
- Fix Oxygen display in Kubuntu
- Remove icons?
- Increase icon size?
- Add parameters?
- Remove def, class ?

### CODE REVIEW

**Comment by ccordoba12** [comment]
This needs a rebase now :-) Also waiting badly for it (just saying ;-)

**Comment by goanpeca** [comment]
I want to implement symbol lookup, it is not that hard from what is already in place... but to have perhaps some testing, we could leave symbol lookup for another one?

**Comment by ccordoba12** [comment]
I think there's a lot here, so yes, symbol lookup can wait for another PR.

**Comment by ccordoba12** [comment]
One last thing: is the new file switcher centered on the Editor when shown? Or on the entire application?

**Comment by goanpeca** [comment]
At this moment, editor

**Comment by ccordoba12** [comment]
Mmmh, I think it'd look better to center it on the entire application (like the "Go to line" dialog). But I'm not sure

**Comment by blink1073** [comment]
For what its worth, Sublime centers on the active editor pane.

**Comment by ccordoba12** [comment]
I thought that was the case. The good thing about Sublime is that (by default) it only has two panes: a (small) file explorer to the left and an editor to the right.

But since our editor takes half of the screen, centering the new switcher just on that area seems a bit awkward to me (specially on small screens).

@goanpeca, what do you think?

**Comment by blink1073** [comment]
When you have a two-column view open in Sublime, it is still centered on the active column.  This feels more natural as it lets you know *where* you are taking action.

**Comment by goanpeca** [comment]
I agree with blinkie

**Comment by ccordoba12** [comment]
Ok, let's leave it like that for now :-) We can revisit this decision later.

Then, please rebase to give this a final review ;-)

**Comment by goanpeca** [comment]
So, it was so close that I wanted to give it a try... here is symbol search :-), round 1

file-switcher

**Comment by ccordoba12** [comment]
Cool! Could you explain what's the meaning of `fs@e`? And how users will distinguish between symbols and files?

Two more remarks:
1. Is this ready for review?
2. Could the line number be added next to the symbol?

**Comment by blink1073** [comment]
@goanpeca, shouldn't the @ match the name of the symbol, not the type of symbol?

**Comment by blink1073** [comment]
http://sublime-text-unofficial-documentation.readthedocs.org/en/latest/file_management/file_management.html#supported-operators

**Comment by goanpeca** [comment]
Well... I thought providing the name of the symbol could allow users to search for only classes or only methods/functions... but it is something we can agree on...

So what do you think should be the behaviour? do you think the type (class or def) should be included in the name displayed, but not on the fuzzy search?

**Comment by goanpeca** [comment]
@ccordoba12 almost ready, I found a small bug that I can fix later today.

Sure, I can add the lune number to the mix ;-)

**Comment by goanpeca** [comment]
The behaviour is the following:

1. The user hits `Ctrl+P` and the fileswitcher appears listing all currently opened files.
2. The user can select a file with down or up arrows directly or can type a fuzzy string to filter in the list, in the image I displayed I used `fs` which would produce a filter list, focused on `file_switcher.py`
3. Here the user has 4 choices... either
     - Cancel (hitting escape, and restore everything to the state before calling the switcher)
     - Hit return and goto to the selected file in the current cursor position
     - Type `fs:45` and move the cursor on line 45
     - Type `fs@` which now will display the symbols (classes and functions) of the currently selected list. Note: This is based on the selected file in the previous step.. so if `fs` filtered 2 results, the user has to explicitely select the one for which the symbols are wanted.
     - The user can also fuzzy search the symbols.. so `fs@anyfuzzystring`

Is it clearer now @ccordoba12 ?

Is this how is works on sublime @blink1073?

**Comment by ccordoba12** [comment]
@goanpeca, thanks for the explanation.

My thoughts:
1. I find `fuzzyfile@fuzzysymbol` too complex to digest. I'd prefer to just follow Sublime convention and use `@` to look for symbols in the currently selected file.
2. You can leave this new syntax if you want (as a feature for power users), but please add `@foo` to look for symbols in the current file (in case is not already there).

I know I'm going to sound like a broken record, but that's why I asked you to leave this for a different PR. Now we're discussing about a new feature on top of a big PR, which makes things harder to review and longer to merge. In contrast, look how easy and fast was to merge your PRs to improve the look and feel of different plugins. All of them are already merged ;-)

**Comment by goanpeca** [comment]
I know I know!, this PR was already super large but in the process of implementing the symbol thing, the codebase vastly improved, it was messy before that... that is why I did it.

What you ask is already already the default, if you tap an @ it will look for symbols in the active file.

**Comment by ccordoba12** [comment]
Ok, no biggie if it helped to improve the code ;-)

Then please fix the error you mentioned so I can start reviewing. I'm going to do a very light review because you already reviewed @d1manson's code in his previous PR.

**Comment by goanpeca** [comment]
@ccordoba12, so I actually changed a lot of the original code to make it more clear...

It should be working now for both symbol and file listing.

**Comment by goanpeca** [comment]
**SPOILER**: there seems to be a bug with the Oxygen theme of Kubuntu which does not display the list items at all , it seems to work fine on the rest though

**Comment by goanpeca** [comment]
@d1manson (if you are curious...)

**Comment by ccordoba12** [comment]
What do you mean by the spoiler? Could you post a screenshot please?

**Comment by goanpeca** [comment]
@blink1073 can you give it a test drive?

File: spyderlib/widgets/editor.py [hunk 1]
```diff
@@ -649,20 +532,27 @@ def clone_from(self, other):
...
-        dlg.edit_file.connect(self.set_stack_index)
-        dlg.close_file.connect(self.close_file)
-        dlg.synchronize(self.get_stack_index())
-        dlg.exec_()
-        self.filelist_dlg = None
-
-    def update_filelistdialog(self):
+        if not self.tabs.count():
+            return
+        e = self
```
Comment (ccordoba12) [inline spyderlib/widgets/editor.py]: Please remove this. Just use `self` below.

Comment (goanpeca) [inline spyderlib/widgets/editor.py]: ok

File: spyderlib/widgets/editor.py [hunk 2]
```diff
@@ -1228,8 +1118,6 @@ def close_file(self, index=None, force=False):
                 if index < new_index:
                     new_index -= 1
                 self.set_stack_index(new_index)
-        if self.get_stack_count() == 0:
-            self.sig_new_file[()].emit()
```
Comment (ccordoba12) [inline spyderlib/widgets/editor.py]: Why did you remove this?

File: spyderlib/widgets/file_switcher.py [hunk 3]
```diff
@@ -0,0 +1,572 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © 2013-2015 The Spyder Development Team
+# Copyright © 2015 Daniel Manson (@d1manson)
+# Licensed under the terms of the MIT License
+# (see spyderlib/__init__.py for details)
+
+from __future__ import print_function
```
Comment (ccordoba12) [inline spyderlib/widgets/file_switcher.py]: Please add a blank line below this one

File: spyderlib/widgets/editor.py (hunk 2 above)
Comment (goanpeca) [inline spyderlib/widgets/editor.py]: I didn't... maybe a rebase error?

Comment (goanpeca) [inline spyderlib/widgets/editor.py]: Fixed

File: spyderlib/widgets/file_switcher.py (hunk 3 above)
Comment (goanpeca) [inline spyderlib/widgets/file_switcher.py]: ok

File: spyderlib/widgets/editor.py (hunk 2 above)
Comment (ccordoba12) [inline spyderlib/widgets/editor.py]: Please push your changes :-)

File: spyderlib/widgets/file_switcher.py (hunk 3 above)
Comment (ccordoba12) [inline spyderlib/widgets/file_switcher.py]: This simple change is missing.

Comment (ccordoba12) [inline spyderlib/widgets/file_switcher.py]: No, sorry.

**Comment by ccordoba12** [comment]
I really don't have time to review the new `FileSwitcher` class. I think it was reviewed throughly by @goanpeca and @Nodd in the previous PR, so I have no problem with that.

So let's wait for @blink1073 approval, after a little bit of live testing (given that he's the one who has used Sublime the most :-)

**Comment by blink1073** [comment]
Uh oh:

```python
Traceback (most recent call last):
  File "/Users/ssilvester/workspace/spyder-ide/spyderlib/spyder.py", line 2471, in call_file_switcher
    self.editor.get_current_editorstack().open_fileswitcher_dlg()
  File "/Users/ssilvester/workspace/spyder-ide/spyderlib/widgets/editor.py", line 539, in open_fileswitcher_dlg
    self.fileswitcher_dlg = FileSwitcher(self, self.tabs, self.data)
  File "/Users/ssilvester/workspace/spyder-ide/spyderlib/widgets/file_switcher.py", line 277, in __init__
    self.setup()
  File "/Users/ssilvester/workspace/spyder-ide/spyderlib/widgets/file_switcher.py", line 573, in setup
    self.setup_file_list(filter_text, current_path)
  File "/Users/ssilvester/workspace/spyder-ide/spyderlib/widgets/file_switcher.py", line 459, in setup_file_list
    short_paths = shorten_paths(self.paths, self.save_status)
  File "/Users/ssilvester/workspace/spyder-ide/spyderlib/widgets/file_switcher.py", line 183, in shorten_paths
    recurse_level({i: pl for i, pl in enumerate(path_list) if pl})
  File "/Users/ssilvester/workspace/spyder-ide/spyderlib/widgets/file_switcher.py", line 125, in recurse_level
    sample_toks = level_idx.values()[0]
TypeError: 'dict_values' object does not support indexing
```

**Comment by goanpeca** [comment]
What the f*ck?

Just by pressing Ctrl+P ? or when doing something? your setup?

**Comment by goanpeca** [comment]
I cannot reproduce this :-( ...

**Comment by blink1073** [comment]
Are you using Python 2? Here's a diff that makes it work for me on 3.4:

```diff
diff --git a/spyderlib/widgets/file_switcher.py b/spyderlib/widgets/file_switcher.py
index c595127..6eab885 100644
--- a/spyderlib/widgets/file_switcher.py
+++ b/spyderlib/widgets/file_switcher.py
@@ -9,6 +9,7 @@ from __future__ import print_function

 import os
 import os.path as osp
+import six

 from spyderlib.qt.QtGui import (QDialog, QHBoxLayout, QIcon, QLabel, QLineEdit,
                                 QListWidget, QListWidgetItem, QRegExpValidator,
@@ -122,7 +123,7 @@ def shorten_paths(path_list, is_unsaved):

         # Firstly, find the longest common prefix for all in the level
         # s = len of longest common prefix
-        sample_toks = level_idx.values()[0]
+        sample_toks = list(level_idx.values())[0]
         if not sample_toks:
             s = 0
         else:
@@ -157,7 +158,7 @@ def shorten_paths(path_list, is_unsaved):
                         group = prospective_group
                     break
                 # Only keep going if all n still match on the kth token
-                _, sample_toks = next(group.iteritems())
+                _, sample_toks = next(six.iteritems(group))
                 prospective_group = {idx: toks for idx, toks
                                      in group.items()
                                      if toks[k] == sample_toks[k]}
@@ -166,7 +167,7 @@ def shorten_paths(path_list, is_unsaved):
                     k += 1
                 else:
                     break
-            _, sample_toks = next(group.iteritems())
+            _, sample_toks = next(six.iteritems(group))
             if k == 0:
                 short_form = ''
             elif k == 1:
```

**Comment by blink1073** [comment]
Also, there is poor contrast:

screen shot 2015-08-16 at 11 26 56 am

**Comment by blink1073** [comment]
Other than those two issues, I like it :stuck_out_tongue_winking_eye:

**Comment by goanpeca** [comment]
Sh#@, I forgot to test on py3.... will make the fix thanks!

**Comment by goanpeca** [comment]
Ok, maybe we need some custom css rules for OSX, it does look funky....

**Comment by goanpeca** [comment]
In spyder we do not use six... we have our own `p3compat.py`, so I guess I will add the iteritems fix there to keep it consistent... (although I would prefer to use `six` PLUS any specific py3compat Qt related stuff....)

**Comment by blink1073** [comment]
Yeah at some point any large project ends up using `six`...

**Comment by goanpeca** [comment]
@blink1073 can you check again?

The contrast issue I will look in a mac this week.

**Comment by goanpeca** [comment]
@ccordoba12 I updated the description with some screenshots...

But it would be nice if you can test it yourself directly

File: spyderlib/widgets/file_switcher.py [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+
+
+# --- Python Outline explorer helpers
+def process_python_symbol_data(oedata):
+    """Returns a list with line number, definition name, fold and token."""
+    symbol_list = []
+    for key in oedata:
+        val = oedata[key]
+        if val and val != 'found_cell_separators':
+            if val.is_class_or_function():
```
Comment (d1manson) [inline spyderlib/widgets/file_switcher.py]: When the file has cells I get the following error here:
`AttributeError: 'bool' object has no attribute 'is_class_or_function'`
If you simply create an empty file and add a `#%%` line you get this error.

**Comment by d1manson** [comment]
@goanpeca thanks for taking over on this!  Symbols lookup is a massive help...if it's not too difficult could you show something a bit more like the tree "path" you see in the outline widget, i.e. `my_method\n[line 99, in my_class]`. Simply trimming whitespace would be nice I think.

**Comment by goanpeca** [comment]
@d1manson, thanks for the heads up, I will see if we can get a more clean display, this is the first iteration, which I think (when I get the cells error fixed) is good for Beta2!, we need people testing this.

Cheers
